node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/nodeadmin/NodeAdminStateUpdater.java

```diff
@@ -145,11 +161,21 @@ public void converge(State wantedState) {
         currentState = wantedState;
     }
 
-    private void adjustNodeAgentsToRunFromNodeRepository() {
+    void adjustNodeAgentsToRunFromNodeRepository() {
         try {
-            final List<NodeSpec> containersToRun = nodeRepository.getNodes(hostHostname);
-            nodeAdmin.refreshContainersToRun(containersToRun);
-        } catch (Exception e) {
+            Map<String, NodeSpec> nodeSpecByHostname = nodeRepository.getNodes(hostHostname).stream()
+                    .collect(Collectors.toMap(NodeSpec::getHostname, Function.identity()));
+            Map<String, Acl> aclByHostname = Optional.of(cachedAclSupplier.get())
+                    .filter(acls -> acls.keySet().containsAll(nodeSpecByHostname.keySet()))
```

Comment:
This can be improved upon, but not sure how. What we want to do is:
If the ACLs were cached in previous or pre-previous tick, and a new container was just allocated, we want to refresh this now and get the right ACL.
If the container was deallocated just after `getNodes()` call, refreshing ACLs is unnecessary.

Reply:
The node agent would easily be able to detect this condition. Add a method on the context to invalidate & get they can call as part of tick? This would unfortunately imply the cache would have to be thread-safe and possibly blocking-free. Or, call each node agent explicitly here and ask if they require a fresh ACL based on the new context.

The not-so-attractive option is to have logic in this class testing for state transitions etc.

Reply:
Well, this is not easy at all... The NodeAgent will not even exist if its the first scenario. It can be done relatively easy by having NASU keep track of hostnames from previous tick, or just ask NodeAdmin for that list and the compare with the current.

If there are no difference, then it's the second scenario (container removed after `getNodes()` call. If they are different, then a new container was just allocated... Not sure if its worth it. It's a rare case and most importantly, the program is correct, just that it sometimes makes an unnecessary request.

Reply:
> The NodeAgent will not even exist if its the first scenario

You must have misunderstood the suggestion. The NodeAgent, in its tick and knowing it has just been created, can issue a invalidate&get request for the ACL through the context. This is not the difficult part - it's coordinating the single cache among the various node agents.

Reply:
That was referring to

> Or, call each node agent explicitly here and ask if they require a fresh ACL based on the new context.

Which I understand as "NASU should ask `NodeAgent`s if it should refresh ACL for their context" and in that case, the `NodeAgent` may not yet exist.

The other suggestion: `NodeAgentContext` is meant to be immutable and consistent within tick. For `NodeAgent` to invalidate anything via Context seems strange..

Reply:
The context would have 2 methods - getPossiblyCachedAcl() and getFreshAclWhichIsExpensive(), and would call whichever is right.

Reply:
Keep in mind that the entire `NodeRepository` itself is cached, so doing a simple `getFreshAclWhichIsExpensive()` may not work. We would need to somehow pass a guaranteed-not-cached client for this particular purpose, otherwise each `NodeAgent` would have to handle the inconsistencies somehow...

Reply:
It's available from ConfigServerTask, so it would just need to be traced through. Not pretty though.